docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 60% of the CPU, and container B requests for 30% of the CPU. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10% of CPU will be distributed to A and B in a 2:1 ratio.
+- Containers will be throttled if they exceed their limit. If limit is unspecified, then the containers can use excess resources when available.
+
+### Incompressible Resource Guarantees
+- For now, we are only supporting memory.
+- Jobs that don’t have a request specified, will have a request of 0.
+- Containers are guaranteed to get the amount of memory they request, if they exceed their memory request, they could be killed (if some other job needs memory), but if they consume fewer resources than requested, they will not be killed.
+- Containers will be killed if they use more memory than their limit.
+
+### Tier Policy
+- *Best-Effort* pods are pods where every container’s request, for both memory and CPU, is 0 or unspecified. These pods are not requesting for any resource guarantees, and will be treated as lowest priority.
+- *Guaranteed* pods are pods where every container, for both memory and CPU, has the same request and limit, and request is non-zero for every container. These pods ask for a well-defined amount of resources and are therefore considered top-priority.
+- All other pods are *Burstable* - middle priority pods that have some form of minimal resource guarantee, but can use more resources when available.
+- A container than specifies a limit of 0 should be considered invalid.
+
+Note that the policy of classifying pods is for the most part orthogonal to the admission policy (described next), and can be easily tweaked.
+
+### Admission Policy
+- Pods will be admitted by Kubelet based on the sum of requests of its containers. The Kubelet will ensure that sum of requests of all containers (over all pods) is within the system’s resource (for both memory and CPU).
+- One big guiding principle is that Best-Effort pods should be “invisible” to all other pods. The main purpose of Best-Effort pods is to utilize resources not used by higher-tier pods. Consequently, Guaranteed and Burstable pods will pre-empt Best-Effort pods if necessary.
+  - Guaranteed pods have a strong guarantee that they will not be pre-empted by any other pods.
+  - Because of implementation challenges, Burstable pods might be pre-empted by Best-Effort pods (but the Kubelet will try to prevent this).
+  - One important instance pre-emption takes into effect is when a node is running the maximum number of pods (currently 40). Suppose that 40 best-effort pods are running on a node, and a guaranteed pod wants to run on the node. The oldest running best-effort pod will be removed and the guaranteed pod will be admitted (if the guaranteed pod’s resources fit in the machine’s capacity).

```

Comment:
It's a flag: https://github.com/GoogleCloudPlatform/kubernetes/blob/9582c7e54ee97a85b7b84a509b572081447ed68d/cmd/kubelet/app/server.go#L248

However, the reason for the max. is to protect Kubelet and Docker. We want to be able to scale to hundreds of pods/containers per node, but we're not there yet. Better to have a max than make it easy to brick nodes.

We also plan to scale it with the number of cores. #12069